### PR TITLE
rumble settings

### LINKED ISSUE #17: Give operator feedback when they attempt to move the arm/wrist into the forbidden zone
Rumble the controller or change the LED colors?

### PR DESCRIPTION
**Pull request by TuskAct2**

### CODE REVIEW

File: src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java [hunk 1]
```diff
@@ -105,6 +114,9 @@ private void configureButtonBindingsManipulator() {
     new POVButton(manipulatorController, Manipulator.intakeConePOV).onTrue(new SetArmWristGoalPreset(GoalPos.INTAKE, () -> false, isFront, arm));
     new POVButton(manipulatorController, Manipulator.substationPickupPOV).onTrue(new SetArmWristGoalPreset(GoalPos.STORED, isCube, isFront, arm));
     new POVButton(manipulatorController, Manipulator.intakeCubePOV).onTrue(new SetArmWristGoalPreset(GoalPos.STORED, () -> true, isFront, arm));
+    new Trigger(getForbFlag).onTrue(new InstantCommand(() -> {manipulatorController.setRumble(RumbleType.kBothRumble,1);}));
+    new Trigger(getForbFlag).onFalse(new InstantCommand(() -> {manipulatorController.setRumble(RumbleType.kBothRumble,0);}));
```
Comment (ProfessorAtomicManiac) [inline src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java]: Your "getForbFlag" is not letting your code build. BEFORE you push make sure your code BUILDS. I'm not sure whether Codespaces has syntax highlighting but make sure there are no red alerts on your ide.

File: src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java [hunk 2]
```diff
@@ -168,4 +184,7 @@ private double inputProcessing(double value) {
   private Trigger axisTrigger(Joystick stick, Axis axis) {
     return new Trigger(() -> Math.abs(getStickValue(stick, axis)) > MIN_AXIS_TRIGGER_VALUE);
   }
+  private Trigger axisTrigger(XboxController stick, Axis axis) {
+    return new Trigger(() -> Math.abs(getStickValue(stick, axis)) > MIN_AXIS_TRIGGER_VALUE);
+  }
```
Comment (ProfessorAtomicManiac) [inline src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java:235]: Instead of having to create an entire new method for every different controller, use `GenericHID` instead since both the `Joystick` class and `XboxController` class both extend `GenericHID`.

File: src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java [hunk 3]
```diff
@@ -269,7 +274,12 @@ public static Translation2d getWristTipPosition(double armPos, double wristPos)
 
     public boolean wristMovementForbidden(double armPos, double wristPos, double wristVelSign) {
         // If the position is not forbidden, then the movement is not forbidden
-        if(!positionForbidden(armPos, wristPos)) return false;
+        if(positionForbidden(armPos, wristPos)) {
+            forbFlag = true;
+            return false;
+        }
```
Comment (ProfessorAtomicManiac) [inline src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java]: Why did you change this portion of the code? Do you think this is the place in the code where it detects that wrist movement is forbidden?

Comment (ProfessorAtomicManiac) [reply src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java]: Hint: `wristMovementForbidden` does tell you whether wrist movement is forbidden (like it says in the method), but does it detect whether the manipulator's controls are valid?

File: src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java [hunk 4]
```diff
@@ -137,7 +139,10 @@ private void driveWrist(TrapezoidProfile.State state) {
     public void setArmTarget(double targetPos, double targetVel) {
         targetPos = getArmClampedGoal(targetPos);
 
-        if(positionForbidden(targetPos, getWristPos())) return;
+        if(positionForbidden(targetPos, getWristPos())) 
+        
+
+        return;
```
Comment (ProfessorAtomicManiac) [inline src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java]: Is something supposed to go here? Don't add unnecessary spaces (Alt+Shift+F will auto format your code). Also if you think some code needs to be written here for manipulator rumbling, what about wrist?

File: src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java [hunk 5]
```diff
@@ -5,6 +5,7 @@
 package org.carlmontrobotics.robotcode2023;
 
 import static org.carlmontrobotics.robotcode2023.Constants.OI.MIN_AXIS_TRIGGER_VALUE;
+import static org.mockito.Answers.values;
```
Comment (ProfessorAtomicManiac) [inline src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java]: Is this import necessary?

**Review by ProfessorAtomicManiac — CHANGES_REQUESTED** [review]

File: src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java [hunk 6]
```diff
@@ -333,6 +352,7 @@ public static boolean positionForbidden(double armPos, double wristPos) {
         boolean ground = tip.getY() < -ARM_JOINT_TOTAL_HEIGHT;
 
         //return horizontal && vertical || ground;
+        forbFlag = true;
```
Comment (CoolSpy3) [inline src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java]: You might want to also create a version of this in wristMovementForbidden so that the operator gets rumble on both wrist and arm movement. The forbFlag is also never unset, so the feature will currently only work once.

File: src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java [hunk 7]
```diff
@@ -120,6 +129,8 @@ private void configureButtonBindingsManipulator() {
     new POVButton(manipulatorController, Manipulator.shelfPickupPOV).onTrue(new SetArmWristGoalPreset(GoalPos.SHELF, isCube, isFront, arm));
     new POVButton(manipulatorController, Manipulator.intakeConePOV).onTrue(new SetArmWristGoalPreset(GoalPos.INTAKE, () -> false, isFront, arm));
     new POVButton(manipulatorController, Manipulator.substationPickupPOV).onTrue(new SetArmWristGoalPreset(GoalPos.STORED, isCube, isFront, arm));
+    new Trigger(() -> {return arm.getForbFlag();}).onTrue(new InstantCommand(() -> {manipulatorController.setRumble(RumbleType.kBothRumble,1);}))
```
Comment (CoolSpy3) [inline src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java]: I might want to make the rumble amount a constant. Don't let this prevent you from merging tho.

**Review by CoolSpy3 — CHANGES_REQUESTED** [review]

File: src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java [hunk 8]
```diff
@@ -39,10 +42,12 @@
...
 import edu.wpi.first.wpilibj2.command.button.Trigger;
 
-public class RobotContainer {
 
+
+public class RobotContainer {
+  
   public final Joystick driverController = new Joystick(Driver.port);
-  public final Joystick manipulatorController = new Joystick(Manipulator.port);
+  public final XboxController manipulatorController = new XboxController(Manipulator.port);
```
Comment (CoolSpy3) [inline src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java]: It doesn't make sense to have one controller be a `Joystick` and one be an `XboxController`. I thought about this, and I think the best solution would be to define both `driverController` and `manipulatorController` as type `GenericHID` and initialize them both to an `XboxController`.

File: src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java [hunk 9]
```diff
@@ -55,6 +60,8 @@ public class RobotContainer {
   public final DigitalInput[] autoSelectors;
 
   public RobotContainer() {
+    
+
```
Comment (CoolSpy3) [inline src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java]:
```suggestion
```

File: src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java [hunk 10]
```diff
@@ -95,6 +102,7 @@ public RobotContainer() {
       () -> inputProcessing(getStickValue(manipulatorController, Axis.kRightY))
     ));
   }
+
```
Comment (CoolSpy3) [inline src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java]:
```suggestion
```

File: src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java (hunk 7 above)
Comment (CoolSpy3) [reply src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java]: The previous comment was written during competition. Now that we are no longer on a time-crunch. Please move the rumble amount to a constant.

File: src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java [hunk 11]
```diff
@@ -120,6 +128,8 @@ private void configureButtonBindingsManipulator() {
     new POVButton(manipulatorController, Manipulator.shelfPickupPOV).onTrue(new SetArmWristGoalPreset(GoalPos.SHELF, isCube, isFront, arm));
     new POVButton(manipulatorController, Manipulator.intakeConePOV).onTrue(new SetArmWristGoalPreset(GoalPos.INTAKE, () -> false, isFront, arm));
     new POVButton(manipulatorController, Manipulator.substationPickupPOV).onTrue(new SetArmWristGoalPreset(GoalPos.STORED, isCube, isFront, arm));
+    new Trigger(() -> {return arm.getForbFlag();}).onTrue(new InstantCommand(() -> {manipulatorController.setRumble(RumbleType.kBothRumble,1);}))
+                                                  .onFalse(new InstantCommand(() -> {manipulatorController.setRumble(RumbleType.kBothRumble,0);}));
```
Comment (CoolSpy3) [inline src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java]:
```suggestion
    new Trigger(arm::getForbFlag).onTrue(new InstantCommand(() -> manipulatorController.setRumble(RumbleType.kBothRumble,1)))
                                                  .onFalse(new InstantCommand(() -> manipulatorController.setRumble(RumbleType.kBothRumble,0)));
```
Readability

File: src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java [hunk 12]
```diff
@@ -152,10 +162,14 @@ public Command getAutonomousCommand() {
...
 
-  private double getStickValue(Joystick stick, Axis axis) {
+  private double getStickValue(GenericHID stick, Axis axis) {
     return stick.getRawAxis(axis.value) * (axis == Axis.kLeftY || axis == Axis.kRightY ? -1 : 1);
   }
 
+  private double getStickValue(XboxController xbox, Axis axis) {
+    return xbox.getRawAxis(axis.value) * (axis == Axis.kLeftY || axis == Axis.kRightY ? -1 : 1);
+  }
+
```
Comment (CoolSpy3) [inline src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java]:
```suggestion
```
This isn't needed if you have an implementation for `GenericHID`.

File: src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java [hunk 13]
```diff
@@ -1,6 +1,48 @@
...
+import static org.carlmontrobotics.robotcode2023.Constants.Arm.kI;
+import static org.carlmontrobotics.robotcode2023.Constants.Arm.kP;
+import static org.carlmontrobotics.robotcode2023.Constants.Arm.kS;
+import static org.carlmontrobotics.robotcode2023.Constants.Arm.kV;
+import static org.carlmontrobotics.robotcode2023.Constants.Arm.offsetRad;
+import static org.carlmontrobotics.robotcode2023.Constants.Arm.posToleranceRad;
+import static org.carlmontrobotics.robotcode2023.Constants.Arm.rotationToRad;
+import static org.carlmontrobotics.robotcode2023.Constants.Arm.velToleranceRadPSec;
+import static org.carlmontrobotics.robotcode2023.Constants.Arm.wristConstraints;
+import static org.carlmontrobotics.robotcode2023.Constants.Arm.wristMotorPort;
```
Comment (CoolSpy3) [inline src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java]: Just use a `*` import

Comment (CoolSpy3) [reply src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java]: Imo, it helps to limit the number of imports required to use wildcard on Organize Imports in vscode settings. Personally, I set `java.sources.organizeImports.staticStarThreshold` to `2`.

File: src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java [hunk 14]
```diff
@@ -28,7 +70,8 @@
 // Arm angle is measured from horizontal on the intake side of the robot and bounded between -3π/2 and π/2
 // Wrist angle is measured relative to the arm with 0 being parallel to the arm and bounded between -π and π (Center of Mass of Roller)
 public class Arm extends SubsystemBase {
-
+    
+    private static boolean forbFlag;
```
Comment (CoolSpy3) [inline src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:33]: Please use a more descriptive name or add comments.

File: src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java [hunk 15]
```diff
@@ -50,7 +93,8 @@ public class Arm extends SubsystemBase {
 
     // rad, rad/s
     public static TrapezoidProfile.State[] goalState = { new TrapezoidProfile.State(-Math.PI / 2, 0), new TrapezoidProfile.State(0, 0) };
-
+    
+
```
Comment (CoolSpy3) [inline src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java]:
```suggestion

```

File: src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java [hunk 16]
```diff
@@ -174,7 +218,10 @@ private void driveWrist(TrapezoidProfile.State state) {
     public void setArmTarget(double targetPos, double targetVel) {
         targetPos = getArmClampedGoal(targetPos);
 
-        if(positionForbidden(targetPos, getWristPos())) return;
+        if(positionForbidden(targetPos, getWristPos())) 
+        
+
+        return;
```
Comment (CoolSpy3) [inline src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java]:
```suggestion
        if(positionForbidden(targetPos, getWristPos())) return;
```

File: src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java [hunk 17]
```diff
@@ -333,7 +390,8 @@ public static boolean positionForbidden(double armPos, double wristPos) {
         boolean ground = tip.getY() < -ARM_JOINT_TOTAL_HEIGHT;
 
         //return horizontal && vertical || ground;
-        return false;
+        forbFlag = ((horizontal && vertical) || ground);
+        return forbFlag;
```
Comment (CoolSpy3) [inline src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java]:
```suggestion
        return (forbFlag = ((horizontal && vertical) || ground));
```
The intermediary `forbFlag` here is confusing.

File: src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java [hunk 18]
```diff
@@ -316,13 +364,22 @@ public boolean wristMovementForbidden(double armPos, double wristPos, double wri
 
         // If the wrist is inside the drivetrain, fold towards the
         if(horizontal && vertical) {
-            return Math.signum(getWristPos() + ROLLER_COM_CORRECTION_RAD) != Math.signum(wristVelSign);
+            forbFlag = Math.signum(getWristPos() + ROLLER_COM_CORRECTION_RAD) != Math.signum(wristVelSign);
+        }
+
+        if (forbFlag) {
+            return true;
```
Comment (CoolSpy3) [inline src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java]: This changes the logic.

File: src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java
Comment (CoolSpy3) [inline src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:1]: You have multiple locations where you are setting `forbFlag` this could lead to a conflict where different parts of the code set/unset the variable, leading to an incorrect state. You're also setting `forbFlag` in `positionForbidden` and `wristMovementForbidden`. These are utility methods and calls to them should not be assumed to correspond to attempted movement of the arm, so please confine the relevant code to `driveArm` and `driveWrist` (and other explicit drive/periodic methods).

**Review by CoolSpy3 — CHANGES_REQUESTED** [review]

File: src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java [hunk 19]
```diff
@@ -53,6 +58,8 @@ public class RobotContainer {
 
   public final PPRobotPath[] autoPaths;
   public final DigitalInput[] autoSelectors;
+  public static final double rumbleFullPower = 1;
+  public static final double rumbleNoPower = 0;
```
Comment (CoolSpy3) [inline src/main/java/org/carlmontrobotics/robotcode2023/RobotContainer.java]: Move to `Constants`

File: src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java (hunk 17 above)
Comment (CoolSpy3) [reply src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java]: Nevermind, ignore this comment.

File: src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java
Comment (CoolSpy3) [inline src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:1]: `forbFlag` is still being set/unset in multiple places which could lead to an incorrect value as the different sections interact with it.
Ex.
driveArm - Movement forbidden - set forbFlag
driveWrist - Movement not forbidden - unset forbFlag
CommandScheduler - Read forbFlag - Value is false, but arm movement is forbidden.

Comment (CoolSpy3) [inline src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java:1]: My bad. In my previous review, I requested that you confine `forbFlag` calls to any of the explicit drive methods (the methods contained in `//#region Drive Methods`) + periodic and cited `driveArm` and `driveWrist`. I should've pointed you to the `setArmTarget` and `setWristTarget` methods (which are still drive methods by the above definition) because that is where the actual forbidden check is performed.

File: src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java [hunk 20]
```diff
@@ -306,7 +318,8 @@ public static Translation2d getWristTipPosition(double armPos, double wristPos)
 
     public boolean wristMovementForbidden(double armPos, double wristPos, double wristVelSign) {
         // If the position is not forbidden, then the movement is not forbidden
-        if(!positionForbidden(armPos, wristPos)) return false;
+        forbFlag = positionForbidden(armPos, wristPos);
```
Comment (CoolSpy3) [inline src/main/java/org/carlmontrobotics/robotcode2023/subsystems/Arm.java]: I think you meant to revert this in your previous commit.

**Review by CoolSpy3 — CHANGES_REQUESTED** [review]
Looks pretty good, just a couple final comments.

In the previous review batch, I made a couple inline suggestions
```suggestion
- example diff here
+ example diff here
```
For those, you have to press the "Commit Suggestion" button or "Add suggestion to batch" and then commit the batch (I forget what the gui looks like) if you want them to be in the branch. If you do not want to commit the suggestion, please add a comment to it so I know it is an intentional nofix. I've gone ahead and unresolved the suggestions this applies to.

**Review by CoolSpy3 — APPROVED** [review]
Looks good to me! Please fix merge conflicts and test this code on the robot before merging this PR.
